Change display order upe methods

Mayisha opened the pull request:
Fixes #2994

This PR implements the ability to order the payment methods on the Stripe settings page when UPE is enabled.

## Changes proposed in this Pull Request:
- Saving the UPE payment method order in an option `stripe_upe_payment_method_order`. If the option is empty but we have the order of legacy payment methods saved, we copy that order to this option to preserve the previous setting.
- Moved the capability check  to the backend.
- To render the payment methods in the saved order on the checkout page, we are reordering them in `reorder_available_payment_gateways` function similar to [the legacy one](https://github.com/woocommerce/woocommerce-gateway-stripe/blob/4f2b0222221d1af815e991ee4f9d3518110d1c13/includes/class-wc-gateway-stripe.php#L658) 

## Testing instructions
- In `develop` branch enable the new checkout experience.
- On the Stripe settings page, note that there is no `Change display order` button at the top of the payment method list.
- Now enable the legacy checkout experience. You should see a `Change display order` button at the top of the payment method list.
- Change display order of a few methods and save changes.
- Now checkout to this (`task/2994-change-display-order-upe`) branch.
- Enable the new checkout experience and refresh the settings page.
- Confirm that the payment method order previously saved in legacy mode still persists and the additional UPE specific payment methods (BNPL, Wallet etc. which are not available in legacy mode) are placed at the end of the list.
- You should see a `Change display order` button at the top of the payment method list as well.
- Now change the display order of a few methods and save changes. If you refresh the page, the new order should persist.
- Enable a few methods randomly.
- As a shopper, add a product to the cart and go to the shortcode checkout page. Confirm that the available payment methods on the checkout page are following the order saved on the settings page.
- This PR does not cover the block checkout, so the order on the block checkout page will not follow the settings. It will be handled in a separate PR.

Mayisha committed:
show sortable list for UPE

Mayisha committed:
save and show ordered payment methods for UPE

Mayisha committed:
Merge branch 'develop' into task/2994-change-display-order-upe

Mayisha committed:
refactor the helper method

Mayisha committed:
saved enabled payment methods in order

Mayisha committed:
show the upe methods on the checkout page in the same order of the settings page

Mayisha committed:
Merge branch 'develop' into task/2994-change-display-order-upe

Mayisha committed:
add changelog

Mayisha committed:
remove unnecessary code

Mayisha committed:
fix tests

Mayisha committed:
use hardcoded enum

Mayisha committed:
copy legacy order to upe order when first swtched to the new checkout experience

Mayisha committed:
add check

Mayisha committed:
Merge branch 'develop' into task/2994-change-display-order-upe

diegocurbelo commented on includes/admin/class-wc-rest-stripe-settings-controller.php:
```diff
@@ -71,7 +71,7 @@ public function register_routes() {
 						'items'             => [
 							'type' => 'string',
-							'enum' => array_merge( WC_Stripe_Helper::get_upe_settings_available_payment_method_ids( $this->gateway ), WC_Stripe_Helper::get_legacy_available_payment_method_ids() ),
+							'enum' => [ 'card', 'alipay', 'giropay', 'klarna', 'affirm', 'afterpay_clearpay', 'eps', 'bancontact', 'boleto', 'ideal', 'oxxo', 'sepa_debit', 'sepa', 'p24', 'sofort', 'multibanco', 'link', 'wechat_pay', 'cashapp' ],
```
We should get these dynamically from the gateway (, this list should be dynamic (or a constant at the Gateway class level), otherwise each time we add a new payment method we need to add it to a bunch of lists

Mayisha replied:
reverted in https://github.com/woocommerce/woocommerce-gateway-stripe/pull/3184/commits/1996f539d60075d8b2bc880714c4ec82f5ec1617

diegocurbelo commented on includes/admin/class-wc-rest-stripe-settings-controller.php:
```diff
@@ -157,7 +157,7 @@ public function register_routes() {
 						'items'             => [
 							'type' => 'string',
-							'enum' => WC_Stripe_Helper::get_legacy_available_payment_method_ids(),
+							'enum' => [ 'card', 'alipay', 'giropay', 'klarna', 'affirm', 'afterpay_clearpay', 'eps', 'bancontact', 'boleto', 'ideal', 'oxxo', 'sepa_debit', 'sepa', 'p24', 'sofort', 'multibanco', 'link', 'wechat_pay', 'cashapp' ],
```
Same as above

diegocurbelo reviewed (changes requested):
Thanks for working on this @Mayisha!

It tests well in a new environment (like E2E, and a new JN site), but I found some problems in my dev environment; it seems that somehow I ended up with "inconsistent" payment method order settings:

![image](https://github.com/woocommerce/woocommerce-gateway-stripe/assets/407542/88e651ea-b74a-4cd4-82c7-e99e59d07d94)

This is causing the settings page to be empty and generate this JS error in the console:
![Screenshot 2024-06-12 at 5 44 42 PM](https://github.com/woocommerce/woocommerce-gateway-stripe/assets/407542/70b0f17e-07a2-4c64-a57e-b3a7f90a751f)

The problem is the final list returned by `WC_Stripe_Helper:get_upe_available_payment_method_ids()`, some of the payment methods have the `stripe_` prefix, which won't match any here: https://github.com/woocommerce/woocommerce-gateway-stripe/blob/develop/client/payment-methods-map.js

![Screenshot 2024-06-12 at 5 51 47 PM](https://github.com/woocommerce/woocommerce-gateway-stripe/assets/407542/b53f3a99-78b3-467d-9386-85f85644f9dc)

We need to add some defensive code to ignore/remove unexpected payment methods ids from the settings (`$stripe_settings['stripe_upe_payment_method_order']`), or just revert to the default if we detect an invalid value.

Mayisha committed:
Merge branch 'develop' into task/2994-change-display-order-upe

Mayisha committed:
dynamically get gateways as before in enum

Mayisha committed:
fix error
- use correct method name after merging in develop
- copy over the legacy order without stripe_ prefix

Mayisha commented:
Thanks @diegocurbelo!

> The problem is the final list returned by WC_Stripe_Helper:get_upe_available_payment_method_ids(), some of the payment methods have the stripe_ prefix,

That's a very good finding. I made a mistake in this commit https://github.com/woocommerce/woocommerce-gateway-stripe/pull/3184/commits/69b27e7505fde7abd03594523d01ec6dd0ec93df where I am copying over the `stripe_legacy_method_order` to the `stripe_upe_payment_method_order` option. The legacy methods have `stripe_` prefix and I need to remove that while copying as UPE methods are not supposed to have this prefix. :woman_facepalming: 

I have fixed it in https://github.com/woocommerce/woocommerce-gateway-stripe/pull/3184/commits/8b17c046c25cac65c751c4c678aded1334ffbc49. You may need to clear the `stripe_upe_payment_method_order` in your dev environment as it is already corrupted. :grimacing:

diegocurbelo reviewed (approved):
Thanks for fixing that @Mayisha!, it's testing well now 🚢 .

![Screenshot 2024-06-24 at 8 01 08 PM](https://github.com/woocommerce/woocommerce-gateway-stripe/assets/407542/f8c459f8-2398-4f90-80dc-6b84f686e60f)

Mayisha committed:
Merge branch 'develop' into task/2994-change-display-order-upe
